hircluster.c

```diff
@@ -3234,15 +3234,73 @@ int redisClusterAppendFormattedCommand(redisClusterContext *cc, char *cmd,
     slot_num = command_format_by_slot(cc, command, commands);
 
     if (slot_num < 0) {
-        goto error;
+        if (command->type == CMD_REQ_REDIS_MULTI) {
+            // This cmd will not be sent until we get a slot from next cmd
+            if (cc->transaction_slot != TRANSACTION_NONE) {
+                __redisClusterSetError(cc, REDIS_ERR_OTHER,
+                                       "transaction already started");
+                goto error;
+            }
+            cc->transaction_slot = TRANSACTION_STARTED;
+            command->cmd = strdup(cmd); // Keep copy of the command to send
+            goto done;
+        } else if (command->type == CMD_REQ_REDIS_EXEC ||
+                   command->type == CMD_REQ_REDIS_DISCARD) {
+            if (cc->transaction_slot == TRANSACTION_NONE) {
+                __redisClusterSetError(cc, REDIS_ERR_OTHER,
+                                       "transaction not started");
+                goto error;
+            }
+            command->slot_num = cc->transaction_slot;
+            cc->transaction_slot = TRANSACTION_NONE;
+        } else {
+            __redisClusterSetError(
+                cc, REDIS_ERR_OTHER,
+                "No keys in command(must have keys for redis cluster mode)");
```

Comment:
These messages are not beautiful, but legacy... Maybe some other time, we can reformulate them.